runner.go

```diff
@@ -366,6 +391,54 @@ func (r *runner) subscribeConfigDir(botCtx context.Context, bot Bot, configDir s
 	}
 }
 
+func (r *runner) superviseBot(runnerCtx context.Context, botType BotType) (context.Context, func(error)) {
+	botCtx, cancel := context.WithCancel(runnerCtx)
+
+	sendAlert := func(err error) {
+		e := r.alerters.alertAll(runnerCtx, botType, err)
+		if e != nil {
+			log.Errorf("Failed to send alert for %s: %+v", botType, e)
+		}
+	}
+
+	// A function that receives an escalated error from Bot.
+	// If critical error is sent, this cancels Bot context to finish its lifecycle.
+	// Bot itself MUST NOT kill itself, but the Runner does. Beware that Runner takes care of all related components' lifecycle.
+	handleError := func(err error) {
+		switch err.(type) {
+		case *BotNonContinuableError:
+			log.Errorf("Stop unrecoverable bot. BotType: %s. Error: %+v", botType, err)
+			cancel()
+
+			go sendAlert(err)
+
+			log.Infof("Stop supervising bot critical error due to context cancellation: %s.", botType)
+
+		default:
+			if r.alertJudge != nil && r.alertJudge(err) {
```

Comment:
As described in its [document](https://github.com/oklahomer/go-sarah/pull/81/commits/f4deea92b9cdbed0263e16110d82ce6d597df62b#diff-a0afb4d13dc79c28cad59d370ba902caR156), `Bot/Adapter` is encouraged to raise any noteworthy error instead of judging if it can handle the error and keep running by itself. It is the runner's task to make such judgment. This `alertJudge` mechanism is part of such a feature. The document itself mentioned about this with an example of where to place a rate limiter.
As an extension to such usage, this mechanism may be required to make a judgment such as "send alert to administrators and stop `Bot` when `Bot` raises `FooReconnectFailureError`s three times within ten seconds." The current implementation with a `bool` return value cannot satisfy such a requirement. The returning value should be capable of acknowledging if the core should send alert and/or stop `Bot`.

Reply:
Changes in b4ae170 lets registered supervisor tell how to react to the failing Bot.